### PR TITLE
feat: add custom email sender hook

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Adds a custom email sender at an API layer. Hook invocation happens on the API so the abstraction is done at the API level instead of on the `mailer` to avoid having the `mailer` import API.

The custom hook email sender serves as a substitute for the default email sender across all endpoints. It is not possible to selectively enable the hook on certain email flows.

### CODE REVIEW

File: internal/api/mail.go [hunk 1]
```diff
@@ -464,7 +529,32 @@ func (a *API) sendEmailChange(r *http.Request, tx *storage.Connection, u *models
...
+		input := hooks.CustomEmailProviderInput{
+			UserID:          u.ID,
+			Email:           u.Email.String(),
+			OTP:             otpCurrent,
+			EmailActionType: mail.EmailChangeCurrentVerification,
+		}
+		output := hooks.CustomEmailProviderOutput{}
+		err = a.invokeHTTPHook(ctx, r, &input, &output, config.Hook.CustomEmailProvider.URI)
+		currentEmail := u.GetEmail()
+		if config.Mailer.SecureEmailChangeEnabled && currentEmail != "" {
```
Comment (J0) [inline internal/api/mail.go:544]: We can potentially merge the two invocations into one by adding additional parameters though that makes handling on the function slightly more cumbersome as there's a need to check for two different paramers `otp` and `newOPT` or similar.

**Comment by J0** [comment]
Going to add RedirectTo and Data.

File: internal/api/hooks.go [hunk 2]
```diff
@@ -189,6 +189,25 @@ func (a *API) invokeHTTPHook(ctx context.Context, r *http.Request, input, output
 			return err
 		}
 
+		if err := json.Unmarshal(response, hookOutput); err != nil {
+			return internalServerError("Error unmarshaling custom SMS provider hook output.").WithInternalError(err)
+		}
+	case *hooks.CustomEmailProviderInput:
```
Comment (hf) [inline internal/api/hooks.go:195]: Let's call it what it is -- `SendEmail`.

File: internal/api/mail.go [hunk 3]
```diff
@@ -269,24 +271,39 @@ func (a *API) sendConfirmation(r *http.Request, tx *storage.Connection, u *model
...
-		u.ConfirmationToken = oldToken
+	if config.Hook.CustomEmailProvider.Enabled {
+		input := hooks.CustomEmailProviderInput{
+			UserID:          user.ID,
+			Email:           user.Email.String(),
+			OTP:             otp,
+			EmailActionType: mail.SignupVerification,
+			UserMetaData:    user.UserMetaData,
+			RedirectTo:      referrerURL,
+		}
```
Comment (hf) [inline internal/api/mail.go:294]: I would maybe scope the email data one level and the user data another level.

Let's provide users with the full `User` object and under `email_data` the OTP, action types, etc.

File: internal/api/hooks.go [hunk 4]
```diff
@@ -186,6 +186,25 @@ func (a *API) invokeHTTPHook(ctx context.Context, r *http.Request, input, output
...
+		hookOutput, ok := output.(*hooks.SendEmailOutput)
+		if !ok {
+			panic("output should be *hooks.SendEmailOutput")
+		}
+
+		var response []byte
+		var err error
+
+		if response, err = a.runHTTPHook(ctx, r, a.config.Hook.SendEmail, input, output); err != nil {
+			return internalServerError("Error invoking custom email provider hook.").WithInternalError(err)
```
Comment (hf) [inline internal/api/hooks.go:202]: Error message should use "send email hook" wording

File: internal/conf/configuration.go [hunk 5]
```diff
@@ -452,6 +452,7 @@ type HookConfiguration struct {
 	PasswordVerificationAttempt ExtensibilityPointConfiguration `json:"password_verification_attempt" split_words:"true"`
 	CustomAccessToken           ExtensibilityPointConfiguration `json:"custom_access_token" split_words:"true"`
 	CustomSMSProvider           ExtensibilityPointConfiguration `json:"custom_sms_provider" split_words:"true"`
+	SendEmail                   ExtensibilityPointConfiguration `json:"custom_email_provider" split_words:"true"`
```
Comment (hf) [inline internal/conf/configuration.go:455]: Update JSON tag

**Review by hf — APPROVED** [review]
Looks good so far! Please fix the nits to align the wording... we should maybe do the same for the SMS one too while it's not fully released yet

**Comment by J0** [comment]
Ah Apologies should have closed this -  I actually opened a new PR #1512 as it was easier to refactor starting from scratch.

We can adjust the wording from `CustomSMSProvider` to `SendSMS` since it's not merged to infra yet. Do so in a separate PR

**Pull request closed without merging**